Implement a new connection pooling mechanism

shafreenAnfar opened the pull request:
We already have two pooling mechanisms but both have issues.
One has high contention for locks and the other has a connection leak.

Following are the things done.

- Fix the idle connection eviction for inbound and outbound.
- Introduce two new connection pooling techniques. 
  - One for high performance
  - One for cater different types transports needs such as controlling outbound connections. 
- Refactor Sourcehandler, TargetHanlder, ConnectionManager, ChannelUtils.
- Removed using the unnecessary thread pool to send message out. 
- Fix the issue in test server.
- Fix the issue of not cleaning the global connection pool when EventLoops are shutdown.
- Tested with a load test and response assertion tests.
- Fix the issue of not invalidating the connection when connection is closed abruptly.

Re-factored the code but a lot more to do. Doing it gracefully.

Commit:
Implement a new connection pooling mechanism

We already have two pooling mechanisms but both have issues.
One has high contention for locks and the other has a connection leak.

Re-factored the code but a lot more to do. Doing it gracefully.

Commit:
Fix the issue not returning 500 when the back-end is down

chanakaudaya commented:
When testing this PR with the latest (0.89) ballerina code, ServiceChaining sample is not working. I'm getting the below error when running the sample and invoking the service.
```
Zip Code 95999
ballerina.lang.errors:Error, message: failed to send the request through the default pooling
	at ClientConnector.post(<native>:0)
	at ATMLocator.locator(servicechaining/samples/ATMLocatorService.bal:23)

```

Commit:
Remov all the unnecessary locks and SH timeout issue

According to the new thread pool we do not need a single
lock. In addition to that fixed the issue of connection
leak on the SH side.

Commit:
Resolve merge conflicts

shafreenAnfar commented:
Hi Chanaka,

Thanks for reporting the issue. The reason for this issue is client connector not sending the "SourceHanlder" property for the second request. This is fixed the below PR.

https://github.com/ballerinalang/ballerina/pull/2593

Commit:
Fix the code line length issue

Commit:
Fix the bug in global thread pool and comment unwanted code

Commit:
Include both locing and non-locing impls

Commit:
Fix typo in the comment

Commit:
Fix global map cleaning issue and testcase issue

Commit:
Add new pool configurator

Commit:
Add a new method to configure new pools

Commit:
Update the connection pool to have new pools

Commit:
Remove old connection pool and default the new one

shafreenAnfar commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/ClientRequestWorker.java:
```diff
@@ -16,8 +16,6 @@
 
```
Please note that this class is there only as reference. Will remove it in the next refactor.

Commit:
Remove unwated code

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports

Commit:
Refactor the code to re-use

Commit:
Avoid waiting indefinitely when maxactive enabled

Commit:
Fix the issue of not returning the connection

Commit:
Remove unwanted white spaces
